controller/controller_client.go

```diff
@@ -5,284 +5,251 @@
-	// The shipper has no queues connected right now, but once it does, this function can't run until
-	// after the queues have emptied and/or shutdown. We'll presumably have a better idea of how this
-	// will work once we have queues connected here.
-	defer func() {
-		c.log.Debugf("shipper has completed shutdown, stopping")
-		c.shutdownComplete.Done()
-	}()
-	c.shutdownComplete.Add(1)
-
-	_ = unit.UpdateState(client.UnitStateHealthy, "Shipper Running", nil)
-	return runner.Start()
-}
-
-func (c *clientHandler) addUnit(unit *client.Unit) {
-	c.unitsMut.Lock()
-	c.units[unit.ID()] = unit
-	c.unitsMut.Unlock()
-}
-
-func (c *clientHandler) getUnit(ID string) *client.Unit {
-	c.unitsMut.Lock()
-	defer c.unitsMut.Unlock()
-	return c.units[ID]
-
-}
-
-func (c *clientHandler) deleteUnit(unit *client.Unit) {
-	c.unitsMut.Lock()
-	delete(c.units, unit.ID())
-	c.unitsMut.Unlock()
+	handler.grpcServer = grpcserver.NewGRPCServer(handler.outputHandler)
+	return handler
 }
 
-// We'll need to track the client ID that's used for the shipper backend itself.
-func (c *clientHandler) setShipperUnitID(unit *client.Unit) {
-	c.shipperOutputID = unit.ID()
-}
+/*
+//////////
+////////// Shipper/output Handlers
+/////////
+*/
 
-// async stop of the shipper's GRPC server, and anything else it needs to manually shutdown
+// close the runner connection. Called from BounceShipper
 func (c *clientHandler) stopShipper() {
-	c.log.Debugf("Stopping Shipper")
-
-	c.shutdownInit <- struct{}{}
+	err := c.outputHandler.Close()
+	if err != nil {
+		c.log.Errorf("error stopping shipper: %s", err)
+	}
 }
 
-// initialize the startup of the shipper grpc server and backend
 func (c *clientHandler) startShipper(unit *client.Unit) {
-	c.log.Debugf("Starting Shipper")
-	atomic.CompareAndSwapUint32(&c.shipperIsStopping, 1, 0)
-
-	// deciding to omit some of these error checks, as the client update state will only return an error if it has a JSON payload to unmarshall
 	_ = unit.UpdateState(client.UnitStateConfiguring, "reading shipper config", nil)
-
-	// Assuming that if we got here from UnitChangedAdded, we don't need to care about the expected state?
 	_, level, unitConfig := unit.Expected()
+
 	cfg, err := config.ShipperConfigFromUnitConfig(level, unitConfig)
 	if err != nil {
-		c.log.Errorf("error unpacking config from agent: %s", err)
-		_ = unit.UpdateState(client.UnitStateFailed, err.Error(), nil)
+		c.reportError("error configuring shipper", err, unit)
 		return
 	}
 
-	err = logp.Configure(cfg.Log)
+	err = c.outputHandler.Start(cfg)
 	if err != nil {
-		c.log.Errorf("error unpacking config from agent: %s", err)
-		_ = unit.UpdateState(client.UnitStateFailed, err.Error(), nil)
+		c.reportError("error starting output server shipper", err, unit)
 		return
 	}
+	_ = unit.UpdateState(client.UnitStateHealthy, "outputs initialized", nil)
+}
+
+// called when we get a UnitUpdated for the output unit
+func (c *clientHandler) updateShipper(unit *client.Unit) {
```

Comment:
```suggestion
func (c *clientHandler) updateShipperOutput(unit *client.Unit) {
```

The name should reflect that it is specific to output units. Also this function should validate the correct type of unit was passed to it.